### ISSUE TITLE
Error parsing string as enum type for `sdrs_advanced_options` on `r/datastore_cluster`

### ISSUE THREAD

**Issue by coffmant** [hashicorp/terraform-provider-vsphere#1448]
### Terraform Version
1.0.1
### vSphere Provider Version
2.0.2
### Affected Resource(s)
- `vsphere_datastore_cluster`
### Terraform Configuration Files
```hcl

resource "vsphere_datastore_cluster" "Desktop" {
    custom_attributes                      = {}
    datacenter_id                          = vsphere_datacenter.datacenter.moid
    folder                                 = "Storage"
    name                                   = "Desktop"
    sdrs_advanced_options                  = {
        "IgnoreAffinityRulesForMaintenance" = "1"
    }
    sdrs_automation_level                  = "automated"
    sdrs_default_intra_vm_affinity         = true
    sdrs_enabled                           = true
    sdrs_free_space_threshold              = 1
    sdrs_free_space_threshold_mode         = "utilization"
    sdrs_free_space_utilization_difference = 25
    sdrs_io_latency_threshold              = 30
    sdrs_io_load_balance_enabled           = true
    sdrs_io_load_imbalance_threshold       = 50
    sdrs_io_reservable_percent_threshold   = 60
    sdrs_io_reservable_threshold_mode      = "automated"
    sdrs_load_balance_interval             = 480
    sdrs_space_utilization_threshold       = 85
    tags                                   = []
}

```

### Debug Output

https://gist.github.com/coffmant/8c5c22968b7d02e427765b57e1041af2

### Expected Behavior

Space utilization threshold should have been changed from 90% -> 85%

### Actual Behavior

See Gist for full output
Error parsing string as enum type ArrayUpdateOperation

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by tenthirtyam** [comment]
Hi @coffmant , I ran a test of this scenario in my environment today and tracked the issue down to:

```hcl
resource "vsphere_datastore_cluster" "datastore_cluster" {
  #... other config ...
  sdrs_advanced_options                  = {
    "IgnoreAffinityRulesForMaintenance" = "1"
    }
}
```

The following example runs without issue:

```hcl
provider "vsphere" {
  vsphere_server       = var.vsphere_server
  user                 = var.vsphere_username
  password             = var.vsphere_password
  allow_unverified_ssl = var.vsphere_insecure
}

data "vsphere_datacenter" "datacenter" {
  name = var.vsphere_datacenter
}

data "vsphere_host" "vsphere_cluster_hosts" {
  count         = length(var.vsphere_cluster_hosts)
  name          = var.vsphere_cluster_hosts[count.index]
  datacenter_id = data.vsphere_datacenter.datacenter.id
}

resource "vsphere_datastore_cluster" "datastore_cluster" {
  datacenter_id                          = data.vsphere_datacenter.datacenter.id
  name                                   = "remote-nfs-dsc"
  sdrs_enabled                           = true
  sdrs_automation_level                  = "automated"
  sdrs_default_intra_vm_affinity         = true
  sdrs_free_space_threshold              = 1
  sdrs_free_space_threshold_mode         = "utilization"
  sdrs_free_space_utilization_difference = 25
  sdrs_io_latency_threshold              = 30
  sdrs_io_load_balance_enabled           = true
  sdrs_io_load_imbalance_threshold       = 50
  sdrs_io_reservable_percent_threshold   = 60
  sdrs_io_reservable_threshold_mode      = "automated"
  sdrs_load_balance_interval             = 480
  sdrs_space_utilization_threshold       = 85
  sdrs_advanced_options = {}
  folder                                 = "nfs"
}

resource "vsphere_nas_datastore" "datastore" {
  for_each             = var.vsphere_datastore_nfs
  type                 = var.vsphere_datastore_type
  name                 = each.value["name"]
  remote_hosts         = each.value["remote_hosts"]
  remote_path          = each.value["remote_path"]
  access_mode          = var.vsphere_datastore_nfs_access_mode
  host_system_ids      = data.vsphere_host.vsphere_cluster_hosts[*].id
  datastore_cluster_id = vsphere_datastore_cluster.datastore_cluster.id
}
```

However, once key/vaules are added to `sdrs_advanced_options` the same error is seen:

```console
❯ terraform apply --auto-approve
vsphere_datastore_cluster.datastore_cluster: Refreshing state... [id=group-p39026]
vsphere_nas_datastore.datastore["datastore1"]: Refreshing state... [id=datastore-39028]
vsphere_nas_datastore.datastore["datastore0"]: Refreshing state... [id=datastore-39029]
vsphere_nas_datastore.datastore["datastore2"]: Refreshing state... [id=datastore-39027]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the
following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # vsphere_datastore_cluster.datastore_cluster will be updated in-place
  ~ resource "vsphere_datastore_cluster" "datastore_cluster" {
        id                                     = "group-p39026"
        name                                   = "remote-nfs-dsc"
      ~ sdrs_advanced_options                  = {
          + "IgnoreAffinityRulesForMaintenance" = "1"
        }
        tags                                   = []
        # (16 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
vsphere_datastore_cluster.datastore_cluster: Modifying... [id=group-p39026]
╷
│ Error: ServerFaultCode: 
│ Error parsing string  as enum type ArrayUpdateOperation
│ 
│ while parsing serialized value of type vim.option.ArrayUpdateSpec.Operation
│ at line 2, column 1044
│ 
│ while parsing property "operation" of static type ArrayUpdateOperation
│ 
│ while parsing serialized DataObject of type vim.storageDrs.OptionSpec
│ at line 2, column 1036
│ 
│ while parsing property "option" of static type ArrayOfStorageDrsOptionSpec
│ 
│ while parsing serialized DataObject of type vim.storageDrs.PodConfigSpec
│ at line 2, column 230
│ 
│ while parsing property "podConfigSpec" of static type StorageDrsPodConfigSpec
│ 
│ while parsing serialized DataObject of type vim.storageDrs.ConfigSpec
│ at line 2, column 224
│ 
│ while parsing call information for method ConfigureStorageDrsForPod_Task
│ at line 2, column 66
│ 
│ while parsing SOAP body
│ at line 2, column 60
│ 
│ while parsing SOAP envelope
│ at line 2, column 0
│ 
│ while parsing HTTP request for method configureStorageDrsForPod
│ on object of type vim.StorageResourceManager
│ at line 1, column 0
│ 
│   with vsphere_datastore_cluster.datastore_cluster,
│   on terraform.tf line 18, in resource "vsphere_datastore_cluster" "datastore_cluster":
│   18: resource "vsphere_datastore_cluster" "datastore_cluster" {
```

The issue occurs with additional advanced options and is not limited to `IgnoreAffinityRulesForMaintenance`.

This does appear to be a bug.

cc @appilon for awareness

**Comment by coffmant** [comment]
@tenthirtyam - Thanks for the investigation!  I must have figured this a long time ago as I have a bunch of these deployed now.  Forgot I even filed this as a bug.

**Comment by tenthirtyam** [comment]
Thanks for the update, @coffmant. Out of curiosity, did you proceed with the datastore clusters w/o the `sdrs_advanced_options` in the Terraform configuration or use another method?

**Ryan Johnson** 
Staff II Solutions Architect, VMware, Inc.
[credly.com/users/tenthirtyam](credly.com/users/tenthirtyam)

**Comment by coffmant** [comment]
I did the manual workaround.  I created the DS cluster without the advanced settings, then updated it simultaneously in both the WebUI and the .tf code, then did a refresh apply (which does work).

So I ended up with the right result but in a very non-TF friendly way.  Obviously this is sub-optimal but we really want full TF coverage on the infra so we're going to work around the sharp edges in the provider.

**Comment by tenthirtyam** [comment]
Thanks again for the update, @coffmant.

Another _option_ could also be to use a PowerShell provisioner to connect and address the gap BMC is PowerCLI or other modules. (I've had to do something very similar for storage policies and tagging for Tanzu.)

Anyway, we have identified it as a bug in the provider for next steps which is a step forward. 😃 

**Ryan Johnson** 
Staff II Solutions Architect, VMware, Inc.
[credly.com/users/tenthirtyam](credly.com/users/tenthirtyam)

**Comment by tenthirtyam** [comment]
@SlimYang has agreed to review this issue.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by zxinyu08** [comment]
Hi @tenthirtyam, I added some debug log for further investigating this issue. The cause should be the spec input to `storagepod.ApplyDRSConfiguration` is not well set. 

I inserted debug log to fully output such spec before https://github.com/hashicorp/terraform-provider-vsphere/blob/main/vsphere/resource_vsphere_datastore_cluster.go#L446. It nests several specs and the innermost layer is type `types.StorageDrsOptionSpec`, and it's exactly generated from field `sdrs_advanced_options` according to https://github.com/hashicorp/terraform-provider-vsphere/blob/main/vsphere/resource_vsphere_datastore_cluster.go#L867

Checked govmomi, `types.StorageDrsOptionSpec` inherits `ArrayUpdateSpec`, I referred to the usage of other spec which also inherits this struct, I noticed ArrayUpdateSpec should also be set explicitly as following shows, otherwise Error parsing string  as enum type ArrayUpdateOperation error will appear.
```
ArrayUpdateSpec: types.ArrayUpdateSpec{
       Operation: types.ArrayUpdateOperationAdd,
},
```

Here's the details, issue reproed -
```
2022-09-01T01:13:05.881-0700 [INFO]  provider.terraform-provider-vsphere: 2022/09/01 01:13:05 [DEBUG] Applying storage DRS configuration against datastore cluster "": timestamp=2022-09-01T01:13:05.880-0700
2022-09-01T01:13:05.881-0700 [WARN]  unexpected data: terraform-example.com/provider/vsphere:stdout="zxinyu-debug: [{ArrayUpdateSpec:{DynamicData:{} Operation: RemoveKey:<nil>} Option:0xc00028a5a0}]  <- no Operation defined
zxinyu-debug: &{DynamicData:{} Key:IgnoreAffinityRulesForMaintenance Value:1}
zxinyu-debug-govmomi-spec: {DynamicData:{} PodConfigSpec:0xc000528300 VmConfigSpec:[]}"
2022-09-01T01:13:05.889-0700 [WARN]  unexpected data: terraform-example.com/provider/vsphere:stdout="zxinyu-debug-govmomi-res: <nil>"      <- API call failed, no task returned
2022-09-01T01:13:05.889-0700 [WARN]  unexpected data: terraform-example.com/provider/vsphere:stdout="zxinyu-debug-govmomi-err: ServerFaultCode:
Error parsing string  as enum type ArrayUpdateOperation

while parsing serialized value of type vim.option.ArrayUpdateSpec.Operation
at line 2, column 1041

while parsing property "operation" of static type ArrayUpdateOperation

while parsing serialized DataObject of type vim.storageDrs.OptionSpec
at line 2, column 1033

while parsing property "option" of static type ArrayOfStorageDrsOptionSpec
```

After fix patched -
```
2022-09-01T00:52:44.488-0700 [WARN]  unexpected data: terraform-example.com/provider/vsphere:stdout="zxinyu-debug: [{ArrayUpdateSpec:{DynamicData:{} Operation:add RemoveKey:<nil>} Option:0xc0006c2460}]  <- Operation set as add, since we need a full set of all sdrs_advanced_options
zxinyu-debug: &{DynamicData:{} Key:IgnoreAffinityRulesForMaintenance Value:1}"
2022-09-01T00:52:44.488-0700 [WARN]  unexpected data: terraform-example.com/provider/vsphere:stdout="zxinyu-debug-govmomi-spec: {DynamicData:{} PodConfigSpec:0xc000587880 VmConfigSpec:[]}"
2022-09-01T00:52:44.501-0700 [WARN]  unexpected data: terraform-example.com/provider/vsphere:stdout="zxinyu-debug-govmomi-res: &{Returnval:Task:task-274}  <- API call succeeded, return vsphere task id
zxinyu-debug-govmomi-err: <nil>"
```
Seen from above logs, besides logging spec, I also output the res and err of `methods.ConfigureStorageDrsForPod_Task` in govmomi to verify if fix works.

**Comment by tenthirtyam** [comment]
Thank you, Xinyu - _this is a wonderful analysis!_

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by tenthirtyam** [comment]
@coffmant - you'll see this resolved in the next release (v2.3.0).

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by coffmant** [comment]
Thank you
